src/dpl/src/Place.cpp

```diff
@@ -932,6 +933,112 @@ bool Opendp::checkRegionOverlap(const Cell* cell,
   // be fully contained by the cell's bounding box.
   return result.empty();
 }
+namespace cell_edges {
+Rect transformEdgeRect(const Rect& edge_rect,
+                       const Cell* cell,
+                       const DbuX x,
+                       const DbuY y,
+                       const odb::dbOrientType& orient)
+{
+  Rect bbox;
+  cell->db_inst_->getMaster()->getPlacementBoundary(bbox);
+  odb::dbTransform transform(orient);
+  transform.apply(bbox);
+  Point offset(x.v - bbox.xMin(), y.v - bbox.yMin());
+  transform.setOffset(offset);
+  Rect result(edge_rect);
+  transform.apply(result);
+  return result;
+}
+
+};  // namespace cell_edges
+bool Opendp::checkEdgeSpacing(const Cell* cell,
+                              const GridX x,
+                              const GridY y,
+                              const odb::dbOrientType& orient) const
+{
+  auto spacing_rules = db_->getTech()->getCellEdgeSpacingTable();
+  if (spacing_rules.empty()) {
+    return true;
+  }
+  const auto& master = db_master_map_.at(cell->db_inst_->getMaster());
+  auto x_real = gridToDbu(x, grid_->getSiteWidth());
+  auto y_real = grid_->gridYToDbu(y);
+  for (const auto& edge1 : master.edges_) {
+    for (auto spacing_rule : db_->getTech()->getCellEdgeSpacingTable()) {
+      if (spacing_rule->isOptional() || spacing_rule->isSoft()) {
+        continue;
+      }
+      std::string second_edge_type;
+      if (spacing_rule->getFirstEdgeType() == edge1.getEdgeType()) {
+        second_edge_type = spacing_rule->getSecondEdgeType();
```

Comment:
Having string comparison in this inner loop feels like a efficiency issue.  I think it would be better to intern the edge type strings in odb and have a getEdgeTypeIndex() method.